src/components/DotNavigation/DotNavigation.tsx

```diff
@@ -0,0 +1,400 @@
+    this.state.start + 4 < this.props.length
+      ? this.renderRadioButton({
+          key: 7,
+          checked: false,
+          size: Size.Tiny,
+          hidden: false,
+          onChange: this.moveToEnd,
+          dataHook: DotNavigationDataHooks.EndButton,
+          ariaLabel: `${this.props.ariaLabel} end`,
+        })
+      : this.renderEmptyButton(7);
+
+  renderAdditionalEndButton = () =>
+    this.state.animation === Animation.Back &&
+    this.state.start + 5 < this.props.length
+      ? this.renderRadioButton({
+          key: 8,
+          checked: false,
+          size: Size.Tiny,
+          hidden: false,
+          onChange: null,
+          dataHook: null,
+          ariaLabel: null,
+        })
+      : this.renderEmptyButton(8);
+
+  renderLongList = () => (
+    <div
+      {...styles('radioGroup', {
+        [Animation.Back]: this.state.animation === Animation.Back,
+        [Animation.Forward]: this.state.animation === Animation.Forward,
+      })}
+      key={this.state.longListKey}
+    >
+      {...[
+        this.renderAdditionalStartButton(),
+        this.renderStartButton(),
+        this.renderPreviousButton(),
+        this.renderMiddleButton(0),
+        this.renderMiddleButton(1),
+        this.renderMiddleButton(2),
+        this.renderNextButton(),
+        this.renderEndButton(),
+        this.renderAdditionalEndButton(),
+      ]}
+    </div>
+  );
+
+  renderShortListButton = (index: number) =>
+    this.renderRadioButton({
+      key: index,
+      checked: index === this.state.current,
+      size: Size.Normal,
+      hidden: false,
+      onChange: this.select(index),
+      dataHook: null,
+      ariaLabel: `${this.props.ariaLabel} ${index}`,
+    });
+
+  renderShortList = () =>
+    new Array(this.props.length)
+      .fill(null)
+      .map((el, index) => this.renderShortListButton(index));
+
+  _getDataAttributes = () => {
+    const { showBorder, theme } = this.props;
+    const { current, start, animation } = this.state;
+
+    return {
+      [DotNavigationDataKeys.ShowBorder]: showBorder,
+      [DotNavigationDataKeys.Theme]: theme,
+      [DotNavigationDataKeys.Current]: current,
+      [DotNavigationDataKeys.Start]: start,
+      [DotNavigationDataKeys.Animation]: animation,
+    };
+  };
+
+  render = () => {
+    const { length, showBorder, theme, ...rest } = this.props;
+
+    if (!isNumber(length) || length <= 0) {
+      return null;
+    }
+
+    return (
+      <TPAComponentsConsumer>
+        {() => (
+          <div
+            {...styles(
+              'root',
+              {
+                bordered: showBorder === true,
+                light: theme === Theme.Light,
+              },
+              rest,
+            )}
+            {...this._getDataAttributes()}
+          >
+            {length <= MAX_SHORT_LIST_LENGTH
+              ? this.renderShortList()
+              : this.renderLongList()}
+          </div>
+        )}
+      </TPAComponentsConsumer>
+    );
```

Comment:
A general note about using `...styles(..)`.
We usually prefer using this **only** on the root element.
Stylable can introduce some bugs when using it like this.

If it's used for states that are for sub-component, then it's better we just toggle classes on/off (I think we import `classnames` in the project)

Reply:
I thought that using states is the main stylable feature. Anyway, I replaces them with toggling classes with classnames)

Reply:
That's true, however, we still encounter strange behaviours when we use "nested" states.
Thanks for the change 💪